src/game/gamemain.ts

```diff
@@ -0,0 +1,318 @@
+import { Dispatch, SetStateAction } from 'react';
+import Resources from './resources';
+import Sprite from './sprite';
+
+type TPlayer = {
+  pos: [ number, number ],
+  side: 'left' | 'right',
+  state: 'flight' | 'shift',
+  sprite: Sprite
+}
+
+type TObject = {
+  pos: [ number, number ],
+  sprite: Sprite
+}
+
+export default class GameMain {
+  canvas: HTMLCanvasElement;
+  ctx: CanvasRenderingContext2D;
+  resources: Resources;
+  lastTime: number;
+
+  player: TPlayer;
+
+  walls: TObject[];
+  debris:TObject[];
+
+  isGameOver: boolean;
+  isGamePaused: boolean;
+  gameTime: number;
+  bgPattern: CanvasPattern | null;
+
+  score: number;
+  col: number;
+
+  playerSpeed: number;
+  wallsSpeed: number;
+  debrisSpeed: number;
+
+  setCollisions: Dispatch<SetStateAction<number>>;
+  setScore: Dispatch<SetStateAction<number>>;
+  setGameOverStatus: Dispatch<SetStateAction<boolean>>;
+  setGamePauseStatus: Dispatch<SetStateAction<boolean>>;
+
+  constructor(canvas: HTMLCanvasElement,
+    setCollisions: Dispatch<SetStateAction<number>>,
+    setScore: Dispatch<SetStateAction<number>>,
+    setGameOverStatus: Dispatch<SetStateAction<boolean>>,
+    setGamePauseStatus: Dispatch<SetStateAction<boolean>>) {
+    this.canvas = canvas;
+    this.ctx = this.canvas.getContext('2d') as CanvasRenderingContext2D;
+    this.lastTime = Date.now();
+    this.isGameOver = false;
+    this.isGamePaused = false;
+
+    this.resources = new Resources();
+
+    this.player = {
+      pos: [ 0, 0 ],
+      side: 'left',
+      state: 'flight',
+      sprite: new Sprite('https://sapcey.netlify.app/img/sprites.png', [ 0, 0 ], [ 34, 34 ], 13, [ 0, 1 ])
+    };
+
+    this.walls = [];
+    this.debris = [];
+
+    this.gameTime = 0;
+    this.bgPattern = null;
+
+    this.score = 0;
+    this.col = 0;
+
+    this.playerSpeed = 300;
+    this.wallsSpeed = 100;
+    this.debrisSpeed = 100;
+
+    this.resources.load([
+      'https://sapcey.netlify.app/img/sprites.png',
+      'https://sapcey.netlify.app/img/bg.png'
+    ]);
+    this.resources.onReady(this.init);
+    this.setControls();
+
+    this.setCollisions = setCollisions;
+    this.setScore = setScore;
+    this.setGameOverStatus = setGameOverStatus;
+    this.setGamePauseStatus = setGamePauseStatus;
+
+  }
+
+  setControls(): void {
+    document.addEventListener('keydown', this.controls);
+  }
+
+  unsetControlsandSubscriptions(): void {
+    document.removeEventListener('keydown', this.controls);
+    this.setCollisions(0);
+    this.setScore(0);
+    this.setGameOverStatus(false);
+    this.setGamePauseStatus(false);
```

Comment:
🧐
Люблю отбивать разные по сути блоки пустой строкой
```suggestion
    document.removeEventListener('keydown', this.controls);
    
    this.setCollisions(0);
    this.setScore(0);
    this.setGameOverStatus(false);
    this.setGamePauseStatus(false);
```